File: backend/app/schemas/task.py
"""변화탐지 작업 스키마."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.sheet import CompareType, ObjectCategory

TaskStatus = Literal[
    "pending",
    "running",
    "succeeded",
    "failed",
    "canceled",
]


class Task(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    description: str
    models: list[ObjectCategory]
    compare_type: CompareType
    standard_resource_id: int | None = None
    compare_resource_id: int | None = None

    sheet_codes: list[str] = Field(default_factory=list)
    status: TaskStatus
    progress: int = 0
    progress_message: str | None = None
    progress_stage: str | None = None
    progress_detail: dict[str, Any] | None = None
    progress_updated_at: datetime | None = None
    created_at: datetime
    started_at: datetime | None = None
    finished_at: datetime | None = None
    celery_task_id: str | None = None

    # 실시간 detection 카운트 — DB COUNT(detections WHERE task_id=id AND NOT is_deleted).
    # serializer 가 호출 시점에 채움. sheets.total_detections 합산이 아닌 단일 진실 원천.
    detection_count: int = 0


class TaskCreatePayload(BaseModel):
    name: str
    description: str = ""
    models: list[ObjectCategory]
    compare_type: CompareType = "image-image"
    standard_resource_id: int
    compare_resource_id: int
    auto_run: bool = True
    """True (기본): 등록 직후 Celery enqueue. False: 작업 row 만 생성, 추후 수동 시작."""


class TaskUpdatePayload(BaseModel):
    """PATCH /tasks/{id} — 부분 수정. None 인 필드는 변경 없음."""
    name: str | None = None
    description: str | None = None
    standard_resource_id: int | None = None
    compare_resource_id: int | None = None
    models: list[ObjectCategory] | None = None

File: frontend/src/types/task.ts
import type { CompareType, ObjectCategory } from "./sheet";
import type { MultiPolygon, Polygon } from "geojson";

/**
 * 변화탐지 작업 (Task) — 1 작업이 N 도엽을 포함.
 * PROMPTS §1 결정: compare_type 은 'image-image' 단일.
 */
export type TaskStatus =
  | "pending" // 큐 대기
  | "running"
  | "succeeded"
  | "failed"
  | "canceled";

export interface Task {
  id: string;
  name: string;
  description: string;
  models: ObjectCategory[];
  compare_type: CompareType;

  standard_resource_id: number | null;
  compare_resource_id: number | null;
  standard_resource_ids: number[];
  compare_resource_ids: number[];

  sheet_codes: string[]; // 작업이 커버하는 도엽
  status: TaskStatus;
  progress: number; // 0~100
  progress_message: string | null;
  progress_stage: string | null;
  progress_detail: Record<string, unknown> | null;
  /** 입력 영상 묶음 양쪽의 합집합을 교차한 실제 처리영역 (EPSG:4326). */
  processing_geometry?: Polygon | MultiPolygon | null;
  processing_area_m2?: number | null;
  progress_updated_at: string | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
  celery_task_id: string | null;

  // 백엔드 단일 진실 원천 — DB 의 활성 detection (NOT is_deleted) 개수.
  // detection import/삭제/추가/복원 시 백엔드 응답에서 갱신된 값을 받아온다.
  detection_count: number;
}

/**
 * 위저드 등록 페이로드 (api/client.ts createTask).
 *
 * auto_run:
 *   true (기본) — 등록 직후 Celery enqueue, 즉시 추론 시작
 *   false — 작업 row 만 생성 (status='pending'), 추후 수동 시작
 */
export interface TaskCreatePayload {
  name: string;
  description: string;
  models: ObjectCategory[];
  compare_type: CompareType;
  standard_resource_id: number;
  compare_resource_id: number;
  standard_resource_ids?: number[];
  compare_resource_ids?: number[];
  auto_run?: boolean;
}
